### ISSUE TITLE
Support for 240p output over either builtin composite or any of the RGB output solutions

### ISSUE THREAD

**Issue by starquake** [libretro/libretro-meta#73]
Now that the Raspberry PI supports 240p output over composite I think we should join effort to get better support for outputting to these kinds of configurations. I hope with some work we can get lakka and retropie to support a simple install that supports composite 240p NTSC output. Which can be used for more advanced users to get output over RGB.

AFAIK currently there are 2 configurations:
1. outputting NTSC over builtin composite - 720x480 getting output to 240p - see https://github.com/raspberrypi/firmware/issues/683
2. outputting over GERT VGA + UMSA or some other form of RGB output (Pi2SCART, Pi2JAMMA) - ???x240 getting output to 240p (variable width, examples: 256 for NES, 320 for Genesis, 1920 to hide scaling issues) - see http://www.retrorgb.com/rpi240p.html

I see common features needed for both outputs:
- Make sure all cores output a height of 240 or 480, no filtering

Specific options for output 1:
- Horizontal filtering to reduce effect caused by non-integer scaling
Specific options for output 2:
- Make sure all cores output a selected width regardless of aspect ratio
- Override menu aspect ratio (RGUI works fine, btw)

Current issues:
- Every core seems to output different resolutions, some do overscan, some don't. Some cores even choose different output resolution based on the game. 

This is my (limited) knowledge on the issue. I would suggest (if I may be so bold) to focus on getting an 'okay' result before getting a perfect result. Any suggestions welcome.

**Comment by mikechi2** [comment]
I think at least the theory is simple and applies to all output methods. We just need a way of consistently (and ideally easily) configuring cores to:

a) Always enforce integer scaling vertically
b) Horizontal scaling should fill the entire width of the screen (for 4:3 displays, add side bars for 16:9), even if the scaling is non-integer

Of course, this causes horizontal distortion, especially if the output resolution is low (e.g., 320 pixels). One solution is to use a higher horizontal resolution (e.g., going from 320 x 240 to 1600 x 240) and do filtered scale. This introduces massive distortions in the console, GUI and everything else. So ideally, there needs to be some mechanism to compensate for that.

**Comment by MrBrax** [comment]
Regardless, if you want a 1:1 320x240 picture in the 720x480 output, you get too much overscan. The clocks/timings are totally wrong for it, so even that's a hindrance.

**Comment by mikechi2** [comment]
@MrBrax That's true, I should've mentioned it. I'll post some pics with my PVM when I get a chance...

I think it's not just overscan. Stretching the image to 720 (using the 720 x 480 240p composite out) actually makes it wider than 4:3 for some reason.

Maybe there needs to be an additional fudge factor that can be tweaked to compensate for overscan while keeping the image centered without needing the user to edit the retroarch.cfg file for every system (which may further vary for every game).

**Comment by MrBrax** [comment]
yes, that's because of the clocks/timings too, since nes/snes at least doesn't run in 4:3, but in 8:7 iirc

**Comment by jayare5** [comment]
This has been glorious :D I have felt thankful for every second I've spent playing in 240p from the Raspberry Pi!

I have all my things perfectly set up already, but I think for it to make sense for all others, theres something else that needs to be considered:
Some kind of menu option that separates SD and HD settings.
So people can switch from an SD TV looking perfectly scaled in 240p, to an HD TV with it filling up the whole screen. In my case, whenever I play on an HDTV from my 240p settings, they appear on a corner of the TV and small.

As for the RGUI, it would be nice if it stayed locked to 240p even when running GameBoy games, because when I do this, the RGUI scales itself down to the GB resolution, which makes it ALMOST unusable!

**Comment by hizzlekizzle** [comment]
I've been using RetroArch at [1920x240](http://filthypants.blogspot.com/2014/11/creating-custom-edid-for-arcade-monitor.html) in [my arcade cabinet](http://filthypants.blogspot.com/2014/11/hooking-pc-to-my-arcade-cabinet.html) for a couple of years now (using a regular Atom PC rather than an RPi). 3840x240 would be better since it's an exact integer of pretty much everything, but I don't notice any distortions at 1920x240 /shrug. Once you set RGUI to match the wide aspect, it looks good and is perfectly usable.

For handheld cores, like Gameboy, the trick is to use the [sgb border shader](https://github.com/hizzlekizzle/glsl-shaders/blob/master/borders/sgb/sgb.glslp) to run at native res without any weird stretching. It looks like this for [GBA](http://2.bp.blogspot.com/-PKJyTjAyZp4/VfiBgX3gkRI/AAAAAAAAC94/qtYdSPS21g0/s1600/IMG_20150716_201832.jpg) and this for [GBC](http://1.bp.blogspot.com/-fYLA7xaGH9A/VfiBgDe1-bI/AAAAAAAAC90/lG4slzsPBM4/s1600/IMG_20150716_201717.jpg). You just modify the glslp file with the appropriate input/output resolutions.

**Comment by mikechi2** [comment]
@hizzlekizzle Can you point me to a link on how to change the RGUI settings? What about the emulationstation front-end? Sorry if it's obvious - I find all the cfg files to be confusing and it might be a good reference for other folks.

**Comment by Ossifex** [comment]
@mikechi2 I think this is what you are looking for:  [https://gist.github.com/Monroe88/dbd3e01252afa5c50690](https://gist.github.com/Monroe88/dbd3e01252afa5c50690)

That is a template for CRT monitors specifically, but it is the same concept of using the ultra wide resolutions and making the RGUI usable.

**Comment by starquake** [comment]
@Mulletizer That looks interesting. I do not have a 480p.cgp however. Where do I get it?

**Comment by starquake** [comment]
Well that was unexpected. I changed my settings based on the ones posted by @Mulletizer and everything seems to be scaling fine now. I used 240p horizontal stripes test where available.
Tested cores
Picodrive, Mednafen PCE Fast, Nestopia, Genesis, Snes9x 2010

I did not try SGB or Gamegear yet, will report back.

These are my current settings:
```
video_crop_overscan = "true"    #SNES scaling is wrong when false. No idea why. Other cores don't care
aspect_ratio_index = "19"
video_fullscreen = "true"
video_fullscreen_x = "720"
video_fullscreen_y = "480"
video_scale_integer = "true"
video_aspect_ratio = "1.333333"
video_aspect_ratio_auto = "false"
custom_viewport_height = "480"
custom_viewport_width = "720"
custom_viewport_x = "0"
custom_viewport_y = "0"
```

**Comment by kevinfishburne** [comment]
I did some 1:1 testing using a Composite A/B switch and my NES and RPi 3 on an SDTV, tweaking the crap out of the settings until the two were nearly identical. What settings I used are in the video. Hopefully someone will find it useful. https://www.youtube.com/watch?v=p8_eKmmN5rA Also nice work everyone. We're on the leading edge of a revolution and it's going to be awesome. :)

[edit] Just read the above posts (starquake and Mulletizer). I'll try those CRT fullscreen settings to see if they're anywhere close to the NES Composite output and report.

**Comment by tomm23** [comment]
@kevinfishburne Looking good man. There is a shader I've used within retroarch to match brightness settings between it and my real SNES. I'm on RetroPie, but it may exist, or could be implemented, in within Lakka as well.

**Comment by mikechi2** [comment]
I got the 1600 x 240 mode working on my component video adapter. The same idea should apply to any of the other DPI based RGB interfaces. Settings are:

hdmi_timings=1600 1 115 150 170 240 1 4 3 15 0 0 0 60 0 32000000 1

![img_7265](https://cloud.githubusercontent.com/assets/23427169/24576796/59e4475e-1677-11e7-9e4d-b46973f47501.JPG)

![img_7264](https://cloud.githubusercontent.com/assets/23427169/24576800/6056dc28-1677-11e7-9b16-e14b88e4ea3d.JPG)

![img_7263](https://cloud.githubusercontent.com/assets/23427169/24576803/64d485d4-1677-11e7-8987-8a8bd971201c.JPG)

![img_7267](https://cloud.githubusercontent.com/assets/23427169/24576806/69f8590a-1677-11e7-98fc-eab2b2548db2.JPG)

Games look awesome. Images fill the entire horizontal width of the screen with no perceptiable scaling errors at 1600 resolution.

But here's the problem:

![img_7266](https://cloud.githubusercontent.com/assets/23427169/24576805/6778485c-1677-11e7-8909-16ef1f68cc60.JPG)

No GUI interfaces are usable :(. Anyone know of a way to stretch the GUI back the other way?

Also maybe I'm tripping but I think the input lag is worse at these settings.

**Comment by mikechi2** [comment]
@Mulletizer Missed your post! OK, time to RGUI!

**Comment by mikoyv3** [comment]
oh wow! all my arcade and consoles work perfectly! Just need to adjust the X Y position and set a per game copnfig to set the right custom_viewport_height (because its either 224 or 240).

maybe a dirty workaround would be to use a centered theme or customized one that takes into consideration the 1600 width.

**Comment by hizzlekizzle** [comment]
After banging my head against HDMI timings unsuccessfully, I switched over to the old hdmi_cvt setting, which worked immediately with my cheap HDMI-to-VGA adapter just by setting hdmi_cvt=1920 240 60. This was hooked up to my XM29+ and will be moved over to my arcade monitor via J-PAC; a 31khz monitor would probably need 1920 480 60 instead with the interlacing shader at 2x.

**Comment by mikechi2** [comment]
@hizzlekizzle yeah I understand, the HDMI timings config is confusing. One thing that would really help is a list of valid pixel clock frequencies... have not been able to dig up the documentation for that.

**Comment by starquake** [comment]
@mikechi2 Look here: https://github.com/raspberrypi/firmware/issues/734

It's a multiple of some value

**Comment by kevinfishburne** [comment]
@mikechi2 So the 1600 ultrawide timings work with your v2 hat? (As in the one you sent me?) I'm about to start testing it and will experiment to see if it works, but thought I'd ask first. [edit] And if it does work, 3840 would be better as it is evenly-divisible by 256.

**Comment by mikechi2** [comment]
Yeah the 1600 settings work. I'm not sure if 3840 x 240 is a valid mode. I was pretty happy with the results at 1600 already. Figure we should try to fix the GUI scaling problem first.

**Comment by kevinfishburne** [comment]
I'll see if RGUI works. If not I'll blind execute the game loading sequence so I can still test until someone resolves the root problem.

**Comment by hizzlekizzle** [comment]
3840 would indeed be ideal but I haven't been able to get it to work. 1920 does, though, and it's an even multiple of 384, which is used by Capcom's CPS1/2/3 boards, if that's a thing you care about.

**Comment by kevinfishburne** [comment]
I care about all the boards. :) I think we're chipping away at a single, elegant solution. If not we should be. The scene has been fractured and arcane for far too long.

**Comment by mikoyv3** [comment]
**what about running these commands before emulationstation loads and after game exits?**
vcgencmd hdmi_timings 320 1 23 25 39 240 1 4 3 15 0 0 0 60 0 6400000 1
tvservice -e "DMT 87"
fbset -depth 8 && fbset -depth 16 -xres 320 -yres 240

**also run these before emulator is loaded**
vcgencmd hdmi_timings 1600 1 115 150 170 240 1 4 3 15 0 0 0 60 0 32000000 1
tvservice -e "DMT 87"
fbset -depth 8 && fbset -depth 16 -xres 1600 -yres 240

**Comment by mikechi2** [comment]
@mikoyv3 Thanks and that sounds promising! Dumb question - where does one insert the commands?

**Comment by mikechi2** [comment]
@hizzlekizzle What's the pixel clock @ 1920?

**Comment by mikoyv3** [comment]
@mikechi2 I found this information on where to insert the commands... 

/opt/retropie/configs/all/runcommand-onstart.sh 
vcgencmd hdmi_timings 1600 1 115 150 170 240 1 4 3 15 0 0 0 60 0 32000000 1
tvservice -e "DMT 87"
fbset -depth 8 && fbset -depth 16 -xres 1600 -yres 240

/opt/retropie/configs/all/runcommand-onend.sh
vcgencmd hdmi_timings 320 1 12 32 44 240 1 6 10 6 0 0 0 60 0 6400000 1
tvservice -e "DMT 87"
fbset -depth 8 && fbset -depth 16 -xres 320 -yres 240

I tried this myself and it seems to be working ! i also changed the hdmi_timings back to 320x240 in /boot/config.txt
I checked the resolution using the command below to be sure.

**FIRST BOOT**
pi@retropie:~ $ tvservice -s
state 0x400000 [LCD], 320x240 @ 0.00Hz, progressive
**DURING GAME**
pi@retropie:~ $ tvservice -s
state 0x400000 [LCD], 1600x240 @ 0.00Hz, progressive
**AFTER GAME EXIT**
pi@retropie:~ $ tvservice -s
state 0x400000 [LCD], 320x240 @ 0.00Hz, progressive

For arcade you have to create a per game config file and set the correct custom_viewport_height (mostly either 224 or 240). Majority are 224 so I made that the default and only created the per game config file (example below) for the handful of games that are 240 height.
/home/pi/RetroPie/roms/arcade/game_name.zip.cfg
video_allow_rotate = "true"
video_rotation = 1
custom_viewport_height = "240"

I just need to figure out how to hide the output on the screen to keep it clean. curious if this works on a per system basis instead of global. Also, given the system and emulator are variables that can be referenced, the runcommand-onstart.sh shell script can be improved to have a dynamic hdmi_timings for each system/emulator. I don't see the need to do it since I have everything working perfectly.

suggestion, increase the emulationstation theme's gameslist font size because 240p makes it barely readable. I personally prefer to use the carbon-nonmeta theme.
/etc/emulationstation/themes/carbon-nometa/carbon_red.xml

:)
---

https://github.com/RetroPie/RetroPie-Setup/wiki/Runcommand

Since 4.0.2 the runcommand can execute a user script before the game launching (runcommand-onstart.sh) and after exiting the emulator (runcommand-onend.sh). Both scripts must be placed at /opt/retropie/configs/all/.

Useful data are passed as arguments to these scripts:

$1 - the system (eg: atari2600, nes, snes, megadrive, fba, etc).
$2 - the emulator (eg: lr-stella, lr-fceumm, lr-picodrive, pifba, etc).
$3 - the full path to the rom file.
$4 - the full command line used to launch the emulator.

**Comment by hizzlekizzle** [comment]
@mikechi2 dunno, actually. I could never get any custom modeline calculations to sync properly, which is why I fell back to the hdmi_cvt setting, since you don't need to calculate the pixel clock by hand with it.

**Comment by netux79** [comment]
I got a bit lost now. Are you still using the composite output to generate the 1600/1920 x 240p width or you are using now the standard RPi HDMI output with an adapter to VGA?
If this is now the HDMI, my question would be... wasn't the new composite 240p good enough for what we were trying to achieve?

**Comment by starquake** [comment]
@netux79 are you asking me?

This issue tries to address all issues related to 240p output which should be similar regardless of width. See the initial description.

**Comment by netux79** [comment]
hi @starquake, yeah I understand it is related to 240p on the RPi. But I think it was initially focused on the new changes to the firmware that allowed to generate 240p thru the builtin composite output of the RPi. And the last comments seem to be using the HDMI output with some kind of adapter to CRT. What I want to know is if this is correct, or if you are still using the builtin composite output. Thanks

**Comment by starquake** [comment]
@netux79 I'm still using the composite on the PI. I hooked up the UMSA this weekend which gives me RGB over SCART but I couldn't stop thinking it's "too sharp"

The problems are the same though. I still need to make sure all consoles line up with full width and 240p height. Some examples. 320x240 (Sega MD), 256x240 (Famicom/NES), 256x192 (Sega Master System after cropping overscan). For the onboard composite this means a width of around 660 (720 minus the overscan you need,which for me is 60) And for the other output stuff this means another arbitrary width.

**Comment by hizzlekizzle** [comment]
@netux79 Yes, my comments are using the HDMI output with an HDMI-to-VGA adapter. Mikechi2's hat is through the GPIO pins. AFAIK, the composite output is still locked to 720x480 non-interlaced.

**Comment by mikoyv3** [comment]
@netux79 sorry I may have hijacked the thread. I am using pi2scart gpio board using vga666 overlay. Its similar to mikechi2's component board.

**Comment by netux79** [comment]
OK, thanks guys for clarifying. So this seems that we still don't have a good-nice alternative yet. If the HDMI to VGA solution looks promising I think I would be able to try it as I have also a cheap chinese adapter  Thanks.

**Comment by starquake** [comment]
I think they're all nice alternatives. And I think the composite one is the closest to the real console.

**Comment by mikoyv3** [comment]
@starquake you can get the vga666 gpio board on ebay for US$7. I read the hdmi to vga adapter introduce a little lag. I am using the pi2scart then converting it to component (yPbPr). I also bought Mike Chi's retrotink component board to simplify my setup.

**Comment by starquake** [comment]
@mikoyv3 yeah as I said I had my UMSA hooked up this weekend. (Using a GERT VGA) I thought it was too sharp. I like the onboard composite 240p NTSC the best because it looks the closest to my Famicom AV.

**Comment by starquake** [comment]
But let's get this issue back on track. I found a config that pretty much causes every system to scale like it should. With that I could probably start thinking about a lakka mod that allows to just hook up a TV to the composite output. The same goes for RetroPie. Not entirely how to approach that. Maybe create a pull request. But what would be the contents? A script that autoconfigures? A general setting? Create an alternative distribution with everything already set up?

The other suggestions I see related to RetroPie are a little bit harder because there are multiple ways of outputting. Am I right?

**Comment by tomm23** [comment]
So is variable timing within composite part of this discussion? no filtering will produce a good signal when it's all scaled across 720.

Also, color burst timing options so that we can avoid ugly vertical rainbow banding as seen in early genesis models?

I'm not sure if these are related to libretro, just pointing out the issues I see. Cool progress nonetheless :)

**Comment by hizzlekizzle** [comment]
@tomm23 In a perfect world, it would be part of the discussion, yes, but that doesn't seem to be in the cards right now, as the firmware is a closed blob.

**Comment by jayare5** [comment]
> closed blob

Ouch. Seems the 240p community is gonna have to find the perfect world to live in and develop its own Raspberry Pi-like computer with its own built-in Composite output that does NOT do the rainbow thing?

**Comment by mikechi2** [comment]
@tomm23  Can you point me to a description of how the colorburst works?

**Comment by tomm23** [comment]
https://en.wikipedia.org/wiki/Colorburst
https://youtu.be/r38nVmxBfvM?t=8m18s
https://youtu.be/mHXDX0MXcyk?t=48s

I can't find where I read how consoles like NES/SNES (and using Genesis 32x attachment) avoided the rainbow banding/fringing issue, but there's a little bit of info:
http://www.chrismcovell.com/gotRGB/rgb_compare.html

**Comment by tommik128** [comment]
> Ouch. Seems the 240p community is gonna have to find the perfect world to live in and develop its own Raspberry Pi-like computer with its own built-in Composite output that does NOT do the rainbow thing?

Wouldn't it be easier to find out the way on Raspberry Pi? But I think it is not going to happen. There are not many people interested in this (my guess is hundreds worldwide maximum) and they will opt for the RGB modules instead. Using composite you will be simply displaying "2 pixels using 5 pixels".

**Comment by starquake** [comment]
@tommik128 I'm actually working on getting the best out of the composite out. When I have found the best config I will try to turn it into a preconfigured rpi image.

It might not be perfect. But I think it's still better (more fun) than playing on a LCD TV with a scanlines shader.

**Comment by sdubrick** [comment]
> I can't find where I read how consoles like NES/SNES (and using Genesis 32x attachment) avoided the rainbow banding/fringing issue,

@tomm23 Check out out these, which have some good information on how the NES dealt with the rainbow banding/dot crawl
http://forums.nesdev.com/viewtopic.php?t=7853
http://wiki.nesdev.com/w/index.php/NTSC_video

**Comment by jayare5** [comment]
It's a hardware issue on the Raspberry Pi, right?
Like, this is because it wasn't meant to do 240p composite from the beginning, i'm assuming since it took so long for a firmware to be made allowing this... right?
Also there's the fact that this requires a shader (at least for me) to look good on composite because without it, things would look too sharp and full of artifacts / dots.

Someone (in the other github thread) made an adapter to allow S-Video to be used, which also had composite... now I wonder, couldn't all composite issues be solved through an adapter like that? So that you don't have to use shaders or get rainbows and instead the adapter does it itself on hardware?

Though that would be an expensive solution unless they made a new revision to the Raspberry Pi boards just for this, which I hope is not as unlikely as I think!
I kinda like to use different shaders to blur things differently though, so the dots and artifacts would have to remain, but if only the thing that's causing the rainbow effect could be done differently...

**Comment by mikechi2** [comment]
The challenge is that systems like the NES generated composite video waveform in precise sync with their framebuffer timing which gives rise to their characteristic NTSC artifacts. In the RPi - there's no control over the video generation circuitry, so they'll be composite artifacts but it won't look the same as the real system.

Probably the best way to simulate that is via shaders.

**Comment by jayare5** [comment]
Shaders are great! But is there a shader that can do something similar to the "composite" option of Nestopia? At least without taking too much CPU.

**Comment by mikechi2** [comment]
Theoretically yes - connect to the TV via RGB, component or S-video, then apply the NTSC artifact shader.

**Comment by jayare5** [comment]
Speaking of which, I've found some shaders that reduce the rainbow effect.
Of course they make the screen more blurry, but it might be possible to remove rainbows without actually making it too blurry, because I've seen other shaders that are REALLY blurry and still have rainbows anyway so it might be unrelated.

I think the one im talking about is the Blur resize 7 or something like that.

**Comment by mikechi2** [comment]
I just tried Blarggs NTSC filter while plugged into my TV via component. The effect is quite convincing. Too bad it caused the emulation to slow down.

**Comment by jayare5** [comment]
I know right? Those filters are great when they work! But... they also dont just work with every game either.
I think I remember the PS1 games not showing anything at all when I had an NTSC filter enabled.

**Comment by MrBrax** [comment]
funnily enough, before this 240p fix came along, i tried applying scanline+ntsc shaders to the original 720x576 output and it looked pretty good on the interlaced signal to the crt 👌

**Comment by manekinekodesu** [comment]
Hello guys, been watching this thread for a while. Yesterday, I've written a small shader to help remove "ripple" effect, visible when scrolling horizontally, which is caused by non-integer horizontal scaling. 

This shader applies selective horizontal blurring to every 5th column of actual pixels, which results in even emulated pixel's width across the screen with negligible loss in sharpness.

And I'm very pleased with how the picture looks now on my CRT TV. I could not tell it apart from integer scaled one, so I decided I'd share it here:
[https://github.com/manekinekodesu/glsl-shaders/blob/master/tv-out-x2.5-pixel-perfect-smoothing-horizontal.glsl](https://github.com/manekinekodesu/glsl-shaders/blob/master/tv-out-x2.5-pixel-perfect-smoothing-horizontal.glsl)

This shader requires the following settings in order to work correctly:
`video_fullscreen_width = "720"`
`custom_viewport_width = "640" # x2.5 scaling of nes/snes resolution`
This settings work  for NES, SNES and GENESIS in 256px mode.

Oh, it also reduces brightness slightly to match original NES composite output.

Edit: link fixed, Added description of the shader

**Comment by starquake** [comment]
@manekinekodesu Very interesting! Will test it later

BTW. When I click on your link it takes me to somewhere else. But I could copy paste the text and that seems to take me to the right page. Here's the link for anyone having the same problem: https://github.com/manekinekodesu/glsl-shaders/blob/master/tv-out-x2.5-pixel-perfect-smoothing-horizontal.glsl

**Comment by starquake** [comment]
@manekinekodesu How do I select the glsl on the pi? It doesn't show up when I browse for it using RetroArch shaders or videofilters. Do I need to compile it somehow?

**Comment by hizzlekizzle** [comment]
@starquake 
Since it's a bare *.glsl file and not a preset, you need to manually tick up the number of shader passes from 0 to 1 and then where it says  "shader # 0" navigate to the glsl file, set the scale and filter (aka, sampling method) as needed and then go up and hit 'apply changes'.

**Comment by manekinekodesu** [comment]
preset file: [https://github.com/manekinekodesu/glsl-shaders/blob/master/tv-out-x2.5-pixel-perfect-smoothing-horizontal.glslp](https://github.com/manekinekodesu/glsl-shaders/blob/master/tv-out-x2.5-pixel-perfect-smoothing-horizontal.glslp)
Needs to be in the same directory the glsl shader is.

**Comment by starquake** [comment]
@manekinekodesu Thanks. Very interested to see what it looks like.

**Comment by Karmeck** [comment]
Regarding gui being un-usable. Rgb-pi seem to have a solution for this using scripts. It's based of recalbox though.

**Comment by vol-2** [comment]
I know this is a bit old, but I think it's worth revisiting as things have changed for the Raspberry Pi 4. 

There is an issue with the GPIO header on the Pi4 in which the RGB lines are prone to interference. As a result, it's more or less impossible to avoid major dot-crawl issues with any onboard digital to VGA conversion via the GPIO.
You can read here: https://www.raspberrypi.org/forums/viewtopic.php?f=91&t=94424&start=175

I found a few modelines that induce less dotcrawl, but they are all close to 320x240 and do not allow for any super-resolutions. Additionally, the dotcrawl effect is extremely sensitive to changes in the pixel clock. Any attempts I made to tweak the resolution with the pixel clock increased screen wobble and crawl to unacceptable levels. This was with the Gert666. With the retrotink, it got even worse. I can only assume this is because the Tink uses more bandwidth on the GPIO, but don't quote me on that.

In light of the poor GPIO VGA image, I believe the current best way to get 240p to your CRT is going to be via an HDMI to VGA adapter that supports 15kHz output.

Some available dongles out there will support various 240p resolutions like 1920x240p.

Using one of these, one can easily output to a CRT. Unfortunately, I'm having issues with support for super resolutions and Lakka with this method. When using a working modeline with a horizontal super-resolution, Lakka centers a 320x240 image in a 1080p frame. Forcing the fullscreen resolution results in a unusable signal.

Would like to get it going if anyone has any suggestions.

**Comment by kevinfishburne** [comment]
I was struggling for a while to get a super resolution working on a RPi4 with the RetroTINK (like a VGA666 but 24-bit/888). I finally got it working with these settings:

```
# Turn off conflicting peripheral overlays using GPIO pins.
dtparam=i2c_arm=off
dtparam=spi=off
dtparam=i2s=off
dtparam=audio=off
enable_uart=0

# Generic 24-bit DPI display.
dtoverlay=dpi24
dpi_output_format=7

# Set default display.
enable_dpi_lcd=1
display_default_lcd=1

# Disable overscan.
disable_overscan=1

# Enable custom dpi timings.
dpi_group=2
dpi_mode=87

# Set custom dpi timings.
dpi_timings=3840 1 109 343 343 240 1 3 5 14 0 0 0 60 0 72925291 1 # 3840x240
#dpi_timings=320 1 9 29 29 224 1 21 3 14 0 0 0 60 0 6077108 1	# 320x224
```

I was connecting to a Sylvania CRT television over S-Video and noticed no dot crawl or artifacts in either of the above modes using both Lakka's November 4th nightly image and the latest Raspbian Buster Lite. Sadly I'm not sure if what fixed the initial issue was the above settings being correct or a firmware update.

Also of note here is that the holy grail of super resolutions, 3840x240, now works. Previously it would fail with a segfault on the RPi3. You may also find this spreadsheet useful for calculating dpi timings:

http://eightvirtues.com/lakka/hat/RPi%20HDMI%20Timings%20Calculator%20and%20Core%20Docs.ods

**Comment by vol-2** [comment]
> I was struggling for a while to get a super resolution working on a RPi4 with the RetroTINK (like a VGA666 but 24-bit/888). I finally got it working with these settings:

Hi Kevin,

Thanks for the reply. That's more or less what I have already done, timings-wise. It does seem to help a bit with the 3840 super res, but it seems to translate the crawl into an overall screen wobble. I suppose not as bad, but the HDMI to VGA solution is even better. It's sharper, less blooming, better contrast, better focus and better colors. After getting it working via Raspbian, it feel like a compromise to go back. lol.
However, as Lakka appears to be incompatible with the HDMI-VGA solution, I guess I'll be using the tink for this. Glad I bought one before Mike Chi discontinued them. :)

One thing I want to add to your settings. The dpi_output_format should be 519. 7 Won't work for many displays because Mike designed the tink ultimate to work a specific way with the falling edge of the pulse. This is supposed to line things up perfectly.

If you want to understand the values, look here: http://retrogamerestore.com/DPI/dpi_out_format_setting.php

Here's Mike's video explanation
https://youtu.be/MAFhVqOoZsc
